File: migrator/progress.py
import json
import os
from dataclasses import dataclass, asdict, field
from typing import List, Optional


@dataclass
class Progress:
    phase: str = "init"
    phase1_start_ts: int = 0
    last_entity_id: Optional[str] = None
    last_entity_ts: int = 0
    watermark_ts: int = 0
    migrated_rows: int = 0
    skipped_rows: int = 0
    started_at: str = ""
    partitioning: str = "MONTHS"
    cast_strings: bool = False
    completed_entities: List[str] = field(default_factory=list)


class ProgressTracker:
    def __init__(self, checkpoint_file: str):
        self.checkpoint_file = checkpoint_file
        self.progress = Progress()

    def load(self) -> bool:
        if not os.path.exists(self.checkpoint_file):
            return False
        with open(self.checkpoint_file) as f:
            data = json.load(f)
        self.progress = Progress(**{k: v for k, v in data.items()
                                    if k in Progress.__dataclass_fields__})
        return True

    def save(self):
        with open(self.checkpoint_file, "w") as f:
            json.dump(asdict(self.progress), f, indent=2)

    def update(self, **kwargs):
        for k, v in kwargs.items():
            setattr(self.progress, k, v)
        self.save()

File: Progress.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace TbMigrator;

public class Progress
{
    [JsonPropertyName("phase")]              public string  Phase              { get; set; } = "init";
    [JsonPropertyName("started_at")]         public string  StartedAt          { get; set; } = "";
    [JsonPropertyName("partitioning")]       public string  Partitioning       { get; set; } = "MONTHS";
    [JsonPropertyName("cast_strings")]       public bool    CastStrings        { get; set; }
    [JsonPropertyName("phase1_start_ts")]    public long    Phase1StartTs      { get; set; }
    [JsonPropertyName("migrated_rows")]      public long    MigratedRows       { get; set; }
    [JsonPropertyName("skipped_rows")]       public long    SkippedRows        { get; set; }
    [JsonPropertyName("watermark_ts")]       public long    WatermarkTs        { get; set; }
    [JsonPropertyName("last_entity_id")]     public string? LastEntityId       { get; set; }
    [JsonPropertyName("completed_entities")] public List<string> CompletedEntities { get; set; } = [];
}

public class ProgressTracker
{
    private static readonly JsonSerializerOptions JsonOpts =
        new() { WriteIndented = true };

    private readonly string _file;
    private readonly object _lock = new();

    public Progress Progress { get; private set; } = new();

    public ProgressTracker(string file) => _file = file;

    public bool Load()
    {
        if (!File.Exists(_file)) return false;
        var json = File.ReadAllText(_file);
        Progress = JsonSerializer.Deserialize<Progress>(json) ?? new Progress();
        return true;
    }

    public void Update(Action<Progress> mutate)
    {
        lock (_lock)
        {
            mutate(Progress);
            File.WriteAllText(_file, JsonSerializer.Serialize(Progress, JsonOpts));
        }
    }
}
